### ISSUE TITLE
Support multiple OIDC providers + config

### ISSUE THREAD

**Issue by m-mohr** [Open-EO/openeo-api#201]
Currently, the endpoint for OpenID Connect `GET /credentials/oidc` redirects to a well-known document with the OIDC metadata. It is not supported for back-ends to support multiple authentication providers, which is limiting. Also, for a good UX it may help to add additional information (especially a name) to the response. 

Instead of a redirect, we could respond with a JSON document like this:
```json
{
  "providers": [
    {
      "authority": "https://accounts.google.com",
      "name": "Google"
    },
    {
      "authority": "https://login.eurac.edu",
      "name": "EURAC"
    },
    {
      "authority": "https://sso.uni-muenster.de",
      "name": "WWU Münster SSO"
    }
  ]
}
```

`/.well-known/openid-configuration` has to be appended to the authority.

We may also need to send more information with it, e.g. config parameters for claims or so. Until now, the Web Editor for example assumes that the user id is in the "sub" claim. Where to get the user id and (optionally) user name from? How to communicate user roles?

Examples for options to consider can be viewed here: https://github.com/IdentityModel/oidc-client-js/wiki#required-settings

Interesting to add could be client_id (not sure), response_type and scope.

What to send as Bearer Token? access_token (response_type = 'token) or id_token (response_type = 'id_token')?

**Comment by m-mohr** [comment]
@aljacob @lforesta Some answers to the questions above:

> We may also need to send more information with it, e.g. config parameters for claims or so. 

We can get scopes from the well known document. Scopes are available in `scopes_supported`. Nevertheless we may want to communicate relevant (and the default?) openEO scopes.

> Until now, the Web Editor for example assumes that the user id is in the "sub" claim. Where to get the user id and (optionally) user name from? How to communicate user roles?

The client should request the information from `GET /me` with the access token. We shouldn't rely on `sub` as it's not standardized. So we need to ensure /me is available and make the workflow consistent between OIDC and Basic.

> What to send as Bearer Token? access_token (response_type = 'token) or id_token (response_type = 'id_token')?

According to https://auth0.com/docs/api-auth/why-use-access-tokens-to-secure-apis it should be the access_token => response_type = token.

Client IDs must be communicated from the Identity provider to the end user and the end user needs to provide it to the client (e.g. textbox in web editor or config file in CLI library). So this needs some "first-time" configuration.

Response from /credentials/oidc would need to be changed to something like:

```json
{
  "providers": [
    {
      "authority": "https://accounts.google.com",
      "name": "Google",
      "scopes": ["earthengine-api"]
    },
    {
      "authority": "https://login.eurac.edu",
      "name": "EURAC",
      "scopes": ["openeo", "internal-data-scope1", "highest-processing-priority-scope"]
    },
    {
      "authority": "https://sso.uni-muenster.de",
      "name": "WWU Münster SSO",
      "scopes": ["ifgi-openeo"]
    }
  ]
}
```

I think I finally have quite a good overview about OIDC and now we can get all this flawlessly running. ;-)

**Comment by m-mohr** [comment]
After some more discussions with @aljacob , it seems we need to send the id_token, not the access_token.

**Comment by soxofaan** [comment]
I'm still bit confused about the client (id) handling.

> Client IDs must be communicated from the Identity provider to the end user 

I'm not sure I understand. This assumes that someone (the backend owner?) has set up a client id in this particular identity provider? 

Related to that: should there also be the option for users to register clients themselves?

At the moment I'm still in the dark how an end user should "get" a client id to use.
I think it would be useful to document or define the scenarios/procedures for doing this  (on global spec level or per backend through e.g. `/credentials/oidc`). Just thinking out loud here about different approaches:
- there is a predefined client id (managed by backend owner) every user can/should use
- there is a predefined client id for simple/demo/readonly/smalldata operations, for advanced usage a dedicated client has to be registered
- end users are expected to register their own client with the identity provider through some API provided by the identity provider
- end users are expected to register their own client with the backend owner through some mechanism (email request or custom API)

**Comment by aljacob** [comment]
I fully agree that we should start now to draft a procedure to follow for client implementations, in order to understand what needs to be done to make it work through oidc and each back-end should also list and inform which identity providers it supports and how a client id can be obtained for a specific back-end.
In most scenarios client applications needs to be registered at the identity provider.
We should not confuse openeo backend driver and identity provider. These are two different things, which are possibly run by the same entity but not necessarily.
Some services like the one offered by google, facebook or github/microsoft, allow some sort of self provisioning for generating the necessary client ids though. In this case however the openeo backend instance also needs to support these identity providers, meaning, they need to implement a validation of identity and authenticity of id-token based on the validation mechanisms provided by the corresponding identity provider.
In the openid workflow client id's are also linked to specific redirect urls, so that after a client has typed in his username and password at the identity providers authentication landing page, it knows where to send the requester back to afterwards.
We as eurac are not intending to become a global identity provider service nor processing service and hence our identity provision is only granted on a per request basis, as in your last example. We basically want to have full control over who we let in and how. For the openEO project and to make access for all partners possible we have however an alternate authentication workflow implemented using basic auth with some fixed defined guest and test accounts.

**Comment by lforesta** [comment]
This was confusing for me too for a long time but this week, with help from @m-mohr and @aljacob , I managed implement the flow as intended.
I'm planning to create a small repo with a fake openeo client/backend and using google as identity provider, this way everyone will be able to test it on their own and/or update/modify the example. I'll write again once the example is available.

**Comment by lforesta** [comment]
I created a simple working example at [https://github.com/Open-EO/openeo-oidc-examples](https://github.com/Open-EO/openeo-oidc-examples), with instructions from the perspective of users, openEO clients and openEO back-ends. I hope this is useful to others.

Maybe we can use that as a starting point to standardize the flow to be used in openEO.
e.g. in the example I use the "Authorization Code" as a Grant Type, in order to be able to get an id_token (instead of an access_token), but I know that @aljacob managed with the "Implicit" Grant Type (for which one only needs to specify the client_id and auth url).

Also, I used Postman there because I don't know how to have a pop-up window in Python where to open the redirect url, maybe someone wants to add that? This will be needed for the Python client anyway (and any CLI-based client).

**Comment by soxofaan** [comment]
> but I know that @aljacob managed with the "Implicit" Grant Type (for which one only needs to specify the client_id and auth url).

OAuth and OpenID is new terrain for me, but from what I've read: the implicit flow is deprecated now and one should use authorization code flow with "PKCE" . For example see here: https://developer.okta.com/blog/2019/05/01/is-the-oauth-implicit-flow-dead   I'm not sure how widely PKCE is supported in providers nowadays, but for new projects/APIs like OpenEO it's probably best to not bother about the implicit flow anymore.

>  I don't know how to have a pop-up window in Python where to open the redirect url, maybe someone wants to add that? This will be needed for the Python client anyway (and any CLI-based client).

I'm working on that in Open-EO/openeo-python-client#82. I have an approach that pops up a webbrowser, even if you are in a CLI environment (using https://docs.python.org/3/library/webbrowser.html). Current implementation: https://github.com/soxofaan/openeo-python-client/blob/EP-2209-openidconnect/openeo/rest/auth/oidc.py   (disclaimer: this is still  work in progress, so quite some things might change based on the testing I am doing now.

**Comment by m-mohr** [comment]
> I'm still bit confused about the client (id) handling.
> 
> > Client IDs must be communicated from the Identity provider to the end user
> 
> I'm not sure I understand. This assumes that someone (the backend owner?) has set up a client id in this particular identity provider?

Yes, a user must request a client id at the provider and assign a redirect url for the client id.


> Related to that: should there also be the option for users to register clients themselves?

You'll probably need that.

> At the moment I'm still in the dark how an end user should "get" a client id to use.
> I think it would be useful to document or define the scenarios/procedures for doing this (on global spec level or per backend through e.g. `/credentials/oidc`).

This is nothing we need to solve within the openEO API. This must be handled by the provider, for example during the registration process (which is also not part of the openEO API).

> * end users are expected to register their own client with the identity provider through some API provided by the identity provider
> * end users are expected to register their own client with the backend owner through some mechanism (email request or custom API)

It should be one of these two options, I guess more likely the first (but may be the second if IP = backend?).


> OAuth and OpenID is new terrain for me, but from what I've read: the implicit flow is deprecated now and one should use authorization code flow with "PKCE" . For example see here: https://developer.okta.com/blog/2019/05/01/is-the-oauth-implicit-flow-dead I'm not sure how widely PKCE is supported in providers nowadays, but for new projects/APIs like OpenEO it's probably best to not bother about the implicit flow anymore.

I don't think we should specify anything regarding flows in the API. It highly depends on the use case, which flow to use. client secrets don't work well in browsers as they can't be hidden, but work well for CLI libraries. Therefore we should let the client implementor decide which workflow to use. The OIDC well-known document will inform which flows are supported.

> > I don't know how to have a pop-up window in Python where to open the redirect url, maybe someone wants to add that? This will be needed for the Python client anyway (and any CLI-based client).

AFAIK the R client (@flahn) just prints out the URL if no browser can be opened. The user can copy&paste the link to his browser and make the appropriate steps there. The client can then spawn up an HTTP server where the authentication details are sent to after the authentication workflow.

**Comment by soxofaan** [comment]
> The client can then spawn up an HTTP server where the authentication details are sent to after the authentication workflow.

FYI: the same feature is now also available in the Python client. see Open-EO/openeo-python-client#84   We want to finetune our own OIDC setup a bit more before sharing a code example/POC however.


> Yes, a user must request a client id at the provider and assign a redirect url for the client id.

This "redirect URL" stuff is the confusing part. Will the end user be able to figure out what to use?

e.g. if the user is using the web editor, then the redirect URL should be something like `http://editor.openeo.org/*` . However when the user wants to use a Jupyter/IPython notebook on their own computer (or another locahost browser based notebook-like system), it has to be something like `http://localhost/*` . If the user wants to use a hosted notebook solution then the redirect url has to correspond to the notebook hosting url (if that is even possible). It is not trivial to explain the what and why of all this.

Setting up an OIDC/OAuth client  is application developer domain and not necessarily very accessible terrain for the general openEO end user I'm afraid.

> This is nothing we need to solve within the openEO API

No indeed, but I think, as I noted above, that some coverage in the openEO documentation is necessary to avoid that end users are going to give up before they even started doing something.

Anyway, we're probably sidetracking/bikeshedding this thread a bit too much and should move the discussion somewhere else?

**Comment by m-mohr** [comment]
> Setting up an OIDC/OAuth client is application developer domain and not necessarily very accessible terrain for the general openEO end user I'm afraid.

Yes, unfortunately we can't do it without redirect URL AFAIK. Otherwise we'd need to drop OIDC/OAuth altogether. The issue here is that we have a n clients and m back-ends, which is a very rare constellation and therefore not really thought of in OIDC/OAuth. 

> Anyway, we're probably sidetracking/bikeshedding this thread a bit too much and should move the discussion somewhere else?

I thought about having a dedicated OIDC call in the next week with everybody implementing OIDC at the moment. This is AFAIK:
* client side: you, @flahn, me
* back-end side: you, @aljacob @lforesta

**Comment by m-mohr** [comment]
@flahn @soxofaan Are you both available for an hour on Tuesday or any of the days afterwards?

**Comment by soxofaan** [comment]
I'm available on Tue 15, Thu 17 and Fri 18

**Comment by m-mohr** [comment]
Fri 14.00 - alternatively Tue 16.00? (Simply answer by using the reactions thumbs up/down)

**Comment by lforesta** [comment]
I can do Friday 18th at 14:00, not Tue (today)

**Comment by m-mohr** [comment]
Great, I'll send more information on the openeo-development mailing list later and remove all comments here regarding the telco discussion so that the issue is a bit less "messy".

**Comment by soxofaan** [comment]
Something that is not yet discussed as far as I know: after a user authenticates against a chosen identity provider, the backend must be informed which provider was picked
For example to be able to resolve the access token (sent as bearer token) to user information (through the userinfo endpoint of the identity provider).
There is so far no mechanism in the openEO API for that I think.

Some simple options (that that can be done without persistent state at backend side):
- send an additional HTTP header describing the chosen provider (in addition to existing HTTP Authorization header with bearer token)
- prefix the access token with provider info when sending as bearer token (e.g. `bearer_token = "{provider_id}:{access_token}"`

**Comment by m-mohr** [comment]
@soxofaan That's indeed a good thought and I'm wondering how other services handle that. Maybe we even need to handle that for Basic Auth. Any preferences?

**Comment by soxofaan** [comment]
Apparently, when the access token is a JWT token, and you decode it, the issuer url is part of the payload under key `iss`. So there is no need for custom means to identify the identity provider.

However it requires that the identify provider issues JWT encoded access tokens, but I think the majority of providers already do this

**Comment by m-mohr** [comment]
AFAIK the EURAC IP (Microsoft) doesn't issue JWT for access_tokens. @aljacob

**Comment by soxofaan** [comment]
Another thing, going back to the original issue (and playing devil's advocate a bit here):

>  It is not supported for back-ends to support multiple authentication providers, which is limiting. 

There is also another solution to this problem: using a single OpenID provider on openEO API level, but one that allows federation to multiple other providers. This avoids the complexity of multiple provider support in the openEO API, client and drivers and leaves it to a specialized service.

That's what we are setting up for VITO backend with Keycloak (https://www.keycloak.org/).
We run our own Keycloak instance and use it as single OpenID provider for the VITO openEO backend.
Keycloak itself however allows to federate authentication to providers like google, github, .... I think it is also possible to integrate non-OpenID systems (e.g. SAML) and expose them with OpenID API.

**Comment by m-mohr** [comment]
I think that solves the issue short-term, but not long-term. Even for Basic/OIDC it could be an issue that the back-end doesn't know whether the Bearer token is coming from Basic or an OIDC provider. And once we want to add other Auth protocols or a second OIDC provider, we need to communicate the information anyway. So I think it would be good to find a mechanism to make multiple Auth procudures working.

**Comment by soxofaan** [comment]
> or you can only support a single auth procedure.

You could even relax that to "or you can only support up to one non-JWT supporting provider in addition to multiple JWT supporting provider"

> Another way is to change the auth workflow altogether.  ...

I'm not sure I understand which problem you are trying to solve. As far as I understand the whole OAuth/OIDC framework, I don't think that it fits the trust relations to let the openEO backend take over control of the access tokens. 
Or do you mean that you want to replace OAuth/OIDC completely with a custom auth flow?

**Comment by m-mohr** [comment]
Decision OIDC telco today:
* Handling the client_id (and secret) is cumbersome, but we can't bake it into the /credentials/oidc endpoint as the redirect_urls are different per client.
* The client_id may be baked into clients for different back-ends, but this feels quite dirty. So in general, users also need a way to take care of it anyway.
* Multiple OIDC providers will be supported, see also #219.
* We don't specify "default" flows (yet?). It depends heavily on the use case/client.
* We don't require access_tokens to be JWT as otherwise we could exclude some OpenID providers.

Did I forget anything? Also, this is a draft and may change whenever problems are encountered.

**Comment by soxofaan** [comment]
@m-mohr  is is already known/decided from which API version this multiple provider feature will be included?  will it be at least 0.5.0?

**Comment by m-mohr** [comment]
@soxofaan Will be included in the next release. ;-) There are currently no plans to release a 0.5.0, but I'm already quite far with the API changes so we could go for a version in-between (1.0-beta or 0.5 or so) in November or so. I'm open to whatever feels most useful to you all. I'm speaking about API only, so excluding any changes to the processes. But processes are very lightly coupled to the API anyway.

**Comment by m-mohr** [comment]
Closed with #225
